### PR TITLE
Support regexp_replace function

### LINKED ISSUE #6115: Support push down of regexp_xxx functions
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

Functions need supported:
- [x] regexp_like
- [x] regexp_instr
- [x] regexp_substr
- [ ] regexp_replace

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #6115 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/OptimizedRegularExpression.inl.h [hunk 1]
```diff
@@ -557,13 +603,94 @@ std::optional<StringRef> OptimizedRegularExpressionImpl<thread_safe>::substrImpl
...
+    size_t byte_offset = byte_pos - 1; // This is a offset for bytes, not utf8
+    StringPieceType expr_sp(subject + byte_offset, subject_size - byte_offset);
+    StringPieceType matched_str;
+
+    while (occur > 0)
+    {
+        bool success = RegexType::FindAndConsume(&expr_sp, *re2, &matched_str);
+        if (!success)
+        {
+            res_data.resize(res_data.size() + subject_size);
```
Comment (windtalker) [inline dbms/src/Common/OptimizedRegularExpression.inl.h:651]: resize to `res_data.size() + subject_size + 1`?

File: dbms/src/Functions/FunctionsRegexp.h [hunk 2]
```diff
@@ -1874,18 +1876,389 @@ class FunctionStringRegexpSubstr : public FunctionStringRegexpBase
...
+
+// The entry to get actual params and execute regexp functions
+#define GET_ACTUAL_PARAMS_AND_EXECUTE() \
+    do                                       \
+    {                                        \
+        GET_EXPR_ACTUAL_PARAM()          \
+    } while (0);
+
+// Implementation of regexp_replace function
+template <typename Impl, typename Name>
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsRegexp.h:1935]: Looks like `Impl` is not used?

File: dbms/src/Common/OptimizedRegularExpression.inl.h (hunk 1 above)
Comment (xzhangxian1008) [reply dbms/src/Common/OptimizedRegularExpression.inl.h:651]: > resize to `res_data.size() + subject_size + 1`?

okk

File: dbms/src/Functions/FunctionsRegexp.h (hunk 2 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsRegexp.h:1935]: > Looks like `Impl` is not used?

I will delete it.

File: dbms/src/Functions/FunctionsRegexp.h [hunk 3]
```diff
@@ -1698,7 +1700,7 @@ class FunctionStringRegexpSubstr : public FunctionStringRegexpBase
 
         auto null_map_col = ColumnUInt8::create();
         typename ColumnUInt8::Container & null_map = null_map_col->getData();
-        null_map.resize(col_size, 1);
+        null_map.resize(col_size);
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsRegexp.h:1703]: Why change this? If `null_map` is not initialized to all `1`s, L1715 should need to assign all the element in `null_map_col` to 1.

File: dbms/src/Functions/FunctionsRegexp.cpp [hunk 4]
```diff
@@ -271,34 +285,32 @@ struct ReplaceRegexpImpl
...
     }
 };
 
 using FunctionTiDBRegexp = FunctionStringRegexp<NameTiDBRegexp>;
 using FunctionRegexpLike = FunctionStringRegexp<NameRegexpLike>;
 using FunctionRegexpInstr = FunctionStringRegexpInstr<NameRegexpInstr>;
 using FunctionRegexpSubstr = FunctionStringRegexpSubstr<NameRegexpSubstr>;
-using FunctionReplaceRegexpOne = FunctionStringReplace<ReplaceRegexpImpl<true>, NameReplaceRegexpOne>;
-using FunctionReplaceRegexpAll = FunctionStringReplace<ReplaceRegexpImpl<false>, NameReplaceRegexpAll>;
+using FunctionRegexpReplace = FunctionStringRegexpReplace<RegexpReplaceImpl, NameRegexpReplace>;
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsRegexp.cpp:306]: Looks like this function only has 1 template argument?

File: dbms/src/Functions/FunctionsRegexp.cpp [hunk 5]
```diff
@@ -23,8 +23,7 @@ namespace DB
 /** Replace all matches of regexp 'needle' to string 'replacement'. 'needle' and 'replacement' are constants.
   * 'replacement' could contain substitutions, for example: '\2-\3-\1'
   */
-template <bool replace_one = false>
-struct ReplaceRegexpImpl
+struct RegexpReplaceImpl
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsRegexp.cpp:26]: Looks like `RegexpReplaceImpl` is not used anymore, I think we can just delete it.

File: dbms/src/Functions/FunctionsRegexp.h [hunk 6]
```diff
@@ -1874,18 +1876,390 @@ class FunctionStringRegexpSubstr : public FunctionStringRegexpBase
...
+        {
+            std::unique_ptr<Regexps::Regexp> regexp;
+            if (col_size > 0)
+            {
+                regexp = memorize<true>(pat_param, match_type_param, collator);
+                if (regexp == nullptr)
+                {
+                    auto null_map_col = ColumnUInt8::create();
+                    typename ColumnUInt8::Container & null_map = null_map_col->getData();
+                    null_map.resize(col_size, 1);
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsRegexp.h:2066]: Should not use `resize(col_size,1)` here.

File: dbms/src/Functions/FunctionsRegexp.h [hunk 7]
```diff
@@ -1030,7 +1032,7 @@ class FunctionStringRegexp : public FunctionStringRegexpBase
                 {
                     auto nullmap_col = ColumnUInt8::create();
                     typename ColumnUInt8::Container & nullmap = nullmap_col->getData();
-                    nullmap.resize(col_size, 1);
+                    nullmap.resize_fill(col_size, 1);
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsRegexp.h:1035]: okk

File: dbms/src/Functions/FunctionsRegexp.h [hunk 8]
```diff
@@ -1721,7 +1728,6 @@ class FunctionStringRegexpSubstr : public FunctionStringRegexpBase
                 {
                     if (expr_param.isNullAt(i) || pos_param.isNullAt(i) || occur_param.isNullAt(i))
                     {
-                        // null_map has been set to 1 in the previous
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsRegexp.h:1724]: Why remove this comment?

**Review by windtalker — APPROVED** [review]
Others LGTM

File: dbms/src/Functions/FunctionsRegexp.h (hunk 8 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsRegexp.h:1724]: > Why remove this comment?

I will add comments

**Review by mengxin9014 — APPROVED** [review]
LGTM

**Pull request merged**